File: cpp/RT_CPP/RT_CPP/database_loader.cpp
#include "database_loader.h"
#include "orthographic.h"
#include "perspective.h"
#include "io/EasyBMP.h"
#include <fstream>
#include <iostream>
#include <algorithm>

using namespace utility::io;

vector<string> split(string s, string delimiter) {
	size_t pos_start = 0, pos_end, delim_len = delimiter.length();
	string token;
	vector<string> res;

	while ((pos_end = s.find(delimiter, pos_start)) != string::npos) {
		token = s.substr(pos_start, pos_end - pos_start);
		pos_start = pos_end + delim_len;
		res.push_back(token);
	}

	res.push_back(s.substr(pos_start));
	return res;
}

material_types database_loader::int_to_enum(int type)
{
	return static_cast<material_types>(type);
}

vector<texture*> database_loader::load_textures(json const& j)
{
	vector<texture*> textures;
	for (json jt : j)
	{
		textures.push_back(load_texture(jt));
	}
	return textures;
}

texture* database_loader::load_texture(string file)
{
	string folder = "scene_files/textures/";
	string full_path = folder + file;
	BMP input;
	input.ReadFromFile(full_path.c_str());
	vector3** pixels = new vector3 * [input.TellHeight()];
	for (int y = 0; y < input.TellHeight(); y++)
	{
		pixels[y] = new vector3[input.TellWidth()];
		for (int x = 0; x < input.TellWidth(); x++)
		{
			pixels[y][x] = { input(x,y)->Red / 255.0f, input(x,y)->Green / 255.0f , input(x,y)->Blue / 255.0f };
		}
	}
	texture* tex = new texture(pixels, input.TellWidth(), input.TellHeight());
	return tex;
}

vector<vector3> database_loader::parse_vector3_array(json const& j)
{
	vector<vector3> vec;
	for (json jv : j)
	{
		vec.push_back(parse_vector3(jv));
	}
	return vec;
}


vector<float> database_loader::parse_float_array(json const& j)
{
	vector<float> vec;
	for (json jv : j)
	{
		vec.push_back(jv);
	}
	return vec;
}

scene_database* database_loader::load(string file)
{
	ifstream in(file.c_str());
	json j;
	in >> j;
	json temp = j["Scene"];
	scene* scn = load_scene(temp);
	temp = j["Camera"];
	camera* cam = load_camera(temp,scn->get_width(),scn->get_height());
	vector<shape*>* shapes = load_shapes(j["Shapes"]);
	vector<material*>* materials = load_materials(j["Materials"]);
	vector<light*>* lights = load_lights(j["Lights"]);
	scene_database* db = new scene_database();
	db->set_camera(cam);
	db->set_scene(scn);
	db->set_shapes(shapes);
	db->set_materials(materials);
	db->set_lights(lights);
	return db;

}

vector3 database_loader::parse_vector3(string vectorStr)
{
	vector<string> vecStr = split(vectorStr, ",");
	vector3 v;
	v.x = atof(vecStr[0].c_str());
	v.y = atof(vecStr[1].c_str());
	v.z = atof(vecStr[2].c_str());
	return v;
}


scene* database_loader::load_scene(json const& j)
{
	scene* s = new scene;
	s->set_height(j["height"]);
	s->set_width(j["width"]);
	s->set_background(parse_vector3(j["background"]));
	s->set_ambient_light(parse_vector3(j["ambientLight"]));
	return s;
}

camera* database_loader::load_camera(json const& j, int width, int height)
{
	bool isOrtho = j["isOrthographic"];

	camera* cam;
	if (isOrtho)
	{
		cam = new orthographic(
			j["farPlane"],
			j["focalLength"],
			j["fieldOfView"],
			parse_vector3(j["upDirection"]),
			parse_vector3(j["target"]),
			parse_vector3(j["position"]),
			width,
			height
		);
	}
	else
	{
		cam = new perspective(
			j["farPlane"],
			j["focalLength"],
			j["fieldOfView"],
			parse_vector3(j["upDirection"]),
			parse_vector3(j["target"]),
			parse_vector3(j["position"]),
			width,
			height
		);
	}

	return cam;
}


vector<shape*>* database_loader::load_shapes(json const& j)
{
	vector<shape*>* shapes = new vector <shape*>;
	json temp = j["Spheres"];
	vector<sphere*>* spheres = load_spheres(temp);
	temp = j["Rectangles"];
	vector<rectangle*>* rectangles = load_rectangles(temp);
	temp = j["Triangles"];
	vector<triangle*>* triangles = load_triangles(temp);
	copy(spheres->begin(), spheres->end(), back_inserter(*shapes));
	copy(rectangles->begin(), rectangles->end(), back_inserter(*shapes));
	copy(triangles->begin(), triangles->end(), back_inserter(*shapes));
	delete spheres;
	delete rectangles;
	delete triangles;
	return shapes;

}


vector<sphere*>* database_loader::load_spheres(json const& j)
{
	vector<sphere*>* spheres = new vector<sphere*>;
	for (json sphJson : j)
	{
		spheres->push_back(load_sphere(sphJson));
	}
	return spheres;
}


sphere* database_loader::load_sphere(json const& j)
{
	sphere* sph = new sphere(j["id"], j["materialId"],
		parse_vector3(j["center"]),
		j["radius"]);
	return sph;
}


vector<rectangle*>* database_loader::load_rectangles(json const& j)
{
	vector<rectangle*>* rectangles = new vector<rectangle*>;
	for (json recJson : j)
	{
		rectangles->push_back(load_rectangle(recJson));
	}
	return rectangles;
}



rectangle* database_loader::load_rectangle(json const& j)
{
	rectangle* rec = new rectangle(j["id"], j["materialId"], parse_vector3( j["vertexA"]),parse_vector3(j["vertexB"]),
		parse_vector3(j["vertexC"]), parse_vector3(j["vertexD"]));
	return rec;
}


vector<triangle*>* database_loader::load_triangles(json const& j)
{
	vector<triangle*>* triangles = new vector<triangle*>;
	for (json triJson : j)
	{
		triangles->push_back(load_triangle(triJson));
	}
	return triangles;
}

triangle* database_loader::load_triangle(json const& j)
{
	triangle* tri = new triangle(j["id"], j["materialId"], parse_vector3(j["vertexA"]), parse_vector3(j["vertexB"]),
		parse_vector3(j["vertexC"]));
	return tri;
}


vector<material*>* database_loader::load_materials(json const& j)
{
	vector<material*>* mats = new vector<material*>;
	vector<color_material*>* color_mats = load_color_materials(j["ColorMaterials"]);
	copy(color_mats->begin(), color_mats->end(), back_inserter(*mats));
	delete color_mats;
	vector<texture_material*>* texture_mats = load_texture_materials(j["TextureMaterials"]);
	copy(texture_mats->begin(), texture_mats->end(), back_inserter(*mats));
	delete texture_mats;
	return mats;
}



color_material* database_loader::load_color_material(json const& j)
{
	color_material* mat = new color_material(j["id"], int_to_enum(j["type"]),
		parse_vector3_array(j["ambientColors"]), parse_vector3_array(j["diffuseColors"]),
		parse_vector3_array(j["specularColors"]), parse_float_array(j["specularities"]),
		parse_float_array(j["reflectivities"]), parse_float_array(j["transparencies"]),
		parse_float_array(j["refractiveIndices"]));
	return mat;
}


vector<texture_material*>* database_loader::load_texture_materials(json const& j)
{
	vector<texture_material*>* cms = new vector<texture_material*>;
	for (json cm : j)
	{
		cms->push_back(load_texture_material(cm));
	}
	return cms;
}

texture_material* database_loader::load_texture_material(json const& j)
{
	texture_material* mat = new texture_material(j["id"], int_to_enum(j["type"]),
		parse_vector3_array(j["ambientColors"]), load_textures(j["diffuseColors"]),
		parse_vector3_array(j["specularColors"]), parse_float_array(j["specularities"]),
		parse_float_array(j["reflectivities"]), parse_float_array(j["transparencies"]),
		parse_float_array(j["refractiveIndices"]));
	return mat;
}

vector<color_material*>* database_loader::load_color_materials(json const& j)
{
	vector<color_material*>* cms = new vector<color_material*>;
	for (json cm : j)
	{
		cms->push_back(load_color_material(cm));
	}
	return cms;
}


vector<light*>* database_loader::load_lights(json const& j)
{
	vector<light*>* lights = new vector<light*>;
	json temp = j["DirectionalLights"];
	vector<directional_light*>* directionals = load_directional_lights(temp);
	temp = j["PointLights"];
	vector<point_light*>* points = load_point_lights(temp);
	temp = j["SpotLights"];
	vector<spot_light*>* spots = load_spot_lights(temp);
	copy(directionals->begin(), directionals->end(), back_inserter(*lights));
	copy(points->begin(), points->end(), back_inserter(*lights));
	copy(spots->begin(), spots->end(), back_inserter(*lights));
	delete directionals;
	delete points;
	delete spots;
	return lights;
}


vector<directional_light*>* database_loader::load_directional_lights(json const& j)
{
	vector<directional_light*>* lights = new vector<directional_light*>;
	for (json jl : j)
	{
		lights->push_back(load_directional_light(jl));
	}
	return lights;
}


directional_light* database_loader::load_directional_light(json const& j)
{
	directional_light* light = new directional_light(j["id"],parse_vector3(j["Color"]),
		parse_vector3(j["Direction"]));
	return light;
}

vector<point_light*>* database_loader::load_point_lights(json const& j)
{
	vector<point_light*>* lights = new vector<point_light*>;
	for (json jl : j)
	{
		lights->push_back(load_point_light(jl));
	}
	return lights;
}


point_light* database_loader::load_point_light(json const& j)
{
	point_light* light = new point_light(j["id"], parse_vector3(j["Color"]), parse_vector3("0,0,0"), parse_vector3(j["Origin"]));
	return light;
}


vector<spot_light*>* database_loader::load_spot_lights(json const& j)
{
	vector<spot_light*>* lights = new vector<spot_light*>;
	for (json jl : j)
	{
		lights->push_back(load_spot_light(jl));
	}
	return lights;
}

spot_light* database_loader::load_spot_light(json const& j)
{
	spot_light* light = new spot_light(j["id"], parse_vector3(j["Color"]), parse_vector3(j["Direction"]), parse_vector3(j["Origin"]),
		j["FallOff"], j["InnerAngle"], j["OuterAngle"]);
	return light;
}

File: Csharp/RT_CS/RT_CS/utility/IO/DatabaseLoader.cs
using Newtonsoft.Json;
using Newtonsoft.Json.Linq;
using RT_CS.Cameras;
using RT_CS.Lights;
using RT_CS.Materials;
using RT_CS.sceneObjects;
using RT_CS.shapes;
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;

namespace RT_CS.utility.IO
{
    public class DatabaseLoader
    {
        public static SceneDatabase Load(string filePath)
        {
            JsonSerializer serializer = new JsonSerializer();
            using StreamReader file = File.OpenText(filePath);
            JObject dbDict = (JObject)
                serializer.Deserialize(file, typeof(JObject));

            SceneDatabase db = new SceneDatabase();
            JObject temp = (JObject)dbDict["Scene"];
            LoadScene(temp, db);
            temp = (JObject)dbDict["Camera"];
            LoadCamera(temp, db);
            db.MyShapes = new List<Shape>();
            JObject shapes = (JObject)dbDict["Shapes"];
            JArray shapesArr = (JArray)shapes["Spheres"];
            LoadSpheres(shapesArr, db);
            shapesArr = (JArray)shapes["Triangles"];
            LoadTriangles(shapesArr, db);
            shapesArr = (JArray)shapes["Rectangles"];
            LoadRectangles(shapesArr, db);
            LoadMaterials((JObject)dbDict["Materials"],db);
            LoadLights((JObject)dbDict["Lights"], db);
            return db;
        }

        private static Vector3 ParseVector3(string vectorStr)
        {
            string[] vectorArr = vectorStr.Split(',');
            Vector3 vector = new Vector3()
            {
                x = Convert.ToSingle(vectorArr[0]),
                y = Convert.ToSingle(vectorArr[1]),
                z = Convert.ToSingle(vectorArr[2])
            };
            return vector;
        }

        private static List<Vector3> ParseVector3s(JArray jArray)
        {
            List<Vector3> vector3s = new List<Vector3>();
            foreach (JValue obj in jArray)
            {
                vector3s.Add(ParseVector3((string)obj));
            }
            return vector3s;

        }


        private static List<float> ParseFloats(JArray jArray)
        {
            List<float> floats = new List<float>();
            foreach (JValue obj in jArray)
            {
                floats.Add((float)obj);
            }
            return floats;
        }
        private static void LoadScene(JObject sceneDict, SceneDatabase db)
        {
            Scene scene = new Scene()
            {
                Height = (int)sceneDict["height"],
                Width = (int)sceneDict["width"],
                Background = ParseVector3((string)sceneDict["background"])
            };
            db.MyScene = scene;
        }

        private static void LoadCamera(JObject cameraDict, SceneDatabase db)
        {
            bool isOrtho = (bool)cameraDict["isOrthographic"];
            Camera camera;
            if (isOrtho)
            {
                camera = new OrthographicCamera((float)cameraDict["focalLength"],
               (float)cameraDict["farPlane"],
               ParseVector3((string)cameraDict["upDirection"]),
               ParseVector3((string)cameraDict["position"]),
               ParseVector3((string)cameraDict["target"]),
               (float)cameraDict["fieldOfView"],
               db.MyScene.Width,
               db.MyScene.Height);
            }
            else
            {
                camera = new PerspectiveCamera((float)cameraDict["focalLength"],
               (float)cameraDict["farPlane"],
               ParseVector3((string)cameraDict["upDirection"]),
               ParseVector3((string)cameraDict["position"]),
               ParseVector3((string)cameraDict["target"]),
               (float)cameraDict["fieldOfView"],
               db.MyScene.Width,
               db.MyScene.Height);
            }
            db.MyCamera = camera;

        }


        private static void LoadSpheres(JArray spheres, SceneDatabase db)
        {
            foreach (var sphere in spheres)
            {
                LoadSphere((JObject)sphere, db);
            }
        }

        private static void LoadRectangles(JArray rectangles, SceneDatabase db)
        {
            foreach (var rectangle in rectangles)
            {
                LoadRectangle((JObject)rectangle, db);
            }
        }


        private static void LoadTriangles(JArray triangles, SceneDatabase db)
        {
            foreach (var triangle in triangles)
            {
                LoadTriangle((JObject)triangle, db);
            }
        }

        private static void LoadSphere(JObject sphereDict, SceneDatabase db)
        {
            Sphere sphere = new Sphere((int)sphereDict["id"], (int)sphereDict["materialId"],
                ParseVector3((string)sphereDict["center"]),
                (float)sphereDict["radius"]);
            db.MyShapes.Add(sphere);
        }

        private static void LoadRectangle(JObject rectDict, SceneDatabase db)
        {
            Rectangle rectangle = new Rectangle((int)rectDict["id"], (int)rectDict["materialId"],
                ParseVector3((string)rectDict["vertexA"]),
                ParseVector3((string)rectDict["vertexB"]),
                ParseVector3((string)rectDict["vertexC"]),
                ParseVector3((string)rectDict["vertexD"]));
            db.MyShapes.Add(rectangle);
        }
        private static void LoadTriangle(JObject triangleDict, SceneDatabase db)
        {
            Triangle triangle = new Triangle((int)triangleDict["id"], (int)triangleDict["materialId"],
                ParseVector3((string)triangleDict["vertexA"]),
                ParseVector3((string)triangleDict["vertexB"]),
                ParseVector3((string)triangleDict["vertexC"]));
            db.MyShapes.Add(triangle);
        }

        private static void LoadMaterials(JObject materials, SceneDatabase db)
        {
            db.MyMaterials = new List<Material>();
            LoadColorMaterials((JArray)materials["ColorMaterials"], db);
        }

        private static void LoadColorMaterials(JArray colorMatArr, SceneDatabase db)
        {
            foreach (JObject colorMat in colorMatArr)
            {
                LoadColorMaterial(colorMat, db);
            }
        }

        private static void LoadColorMaterial(JObject colorMatDict, SceneDatabase db)
        {
            ColorMaterial material = new ColorMaterial((int)colorMatDict["id"],
                ParseFloats((JArray)colorMatDict["specularities"]), (MaterialType)(int)colorMatDict["type"],
                ParseVector3s((JArray)colorMatDict["ambientColors"]), ParseVector3s((JArray)colorMatDict["diffuseColors"]),
                ParseVector3s((JArray)colorMatDict["specularColors"]));
            db.MyMaterials.Add(material);
        }

        private static void LoadLights(JObject lightDict, SceneDatabase db)
        {
            db.MyLights = new List<Light>();
            LoadDirectionalLights((JArray)lightDict["DirectionalLights"], db);
            LoadPointLights((JArray)lightDict["PointLights"], db);
            LoadSpotLights((JArray)lightDict["SpotLights"], db);
        }

        private static void LoadDirectionalLights(JArray lightArr,SceneDatabase db)
        {
            foreach(JObject light in lightArr)
            {
                LoadDirectionalLight(light,db);
            }
        }

        private static void LoadPointLights(JArray lightArr,SceneDatabase db)
        {
            foreach (JObject light in lightArr)
            {
                LoadPointLight(light,db);
            }
        }

        private static void LoadSpotLights(JArray lightArr, SceneDatabase db)
        {
            foreach(JObject light in lightArr)
            {
                LoadSpotLight(light,db);
            }
        }

        private static void LoadDirectionalLight(JObject dirLightDict,SceneDatabase db)
        {
            DirectionalLight light = new DirectionalLight((int)dirLightDict["id"],
                ParseVector3((string)dirLightDict["Direction"]), ParseVector3((string)dirLightDict["Color"]));
            db.MyLights.Add(light);
        }

        private static void LoadPointLight(JObject pointLightDict, SceneDatabase db)
        {
            PointLight light = new PointLight((int)pointLightDict["id"], new Vector3(), ParseVector3((string)pointLightDict["Color"]),
                ParseVector3((string)pointLightDict["Origin"]));
            db.MyLights.Add(light);
        }

        private static void LoadSpotLight(JObject spotLightDict, SceneDatabase db)
        {
            SpotLight light = new SpotLight((int)spotLightDict["id"], ParseVector3((string)spotLightDict["Direction"]),
                ParseVector3((string)spotLightDict["Color"]), ParseVector3((string)spotLightDict["Origin"]),
                (float)spotLightDict["InnerAngle"], (float)spotLightDict["OuterAngle"], (float)spotLightDict["FallOff"]);
            db.MyLights.Add(light);
        }
    }
}
